### PR TITLE
[Merged by Bors] - feat: Port/Data.Finset.Pointwise

### PR DESCRIPTION
**Pull request by arienmalec**
port of data.finset.pointwise

---

- [x] depends on: #1946

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by LukasMias** [comment]
Opened a Zulip thread on some of the `to_additive` failures in this file: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/to_additive.20for.20generated.20lemmas

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#1946
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1946~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by LukasMias** [comment]
`Finset.vsub` apparently used to be a dangerous instance, due to the `β` in `VSub α β` being an `outParam` and hence producing subgoals of the form `[Decidable ?α]`. The usual tricks failed me, but switching around `VSub α β` and `Decidable α` works. Though this feels like something that could break things down the line & might need backporting? Some expert opinion would be appreciated, maybe there's a better fix, too.

File: Mathlib/Data/Finset/Pointwise.lean [hunk 1]
```diff
@@ -0,0 +1,2010 @@
...
+section Mul
+
+variable [DecidableEq α] [DecidableEq β] [Mul α] [Mul β] [MulHomClass F α β] (f : F)
+  {s s₁ s₂ t t₁ t₂ u : Finset α} {a b : α}
+
+/-- The pointwise multiplication of finsets `s * t` and `t` is defined as `{x * y | x ∈ s, y ∈ t}`
+in locale `Pointwise`. -/
+@[to_additive
+      "The pointwise addition of finsets `s + t` is defined as `{x + y | x ∈ s, y ∈ t}` in
+      locale `pointwise`."]
```
Comment (jcommelin) [inline Mathlib/Data/Finset/Pointwise.lean:299]:
```suggestion
      locale `Pointwise`."]
```

File: Mathlib/Data/Finset/Pointwise.lean [hunk 2]
```diff
@@ -0,0 +1,2010 @@
...
+#align finset.subset_add Finset.subset_add
+
+@[to_additive]
+theorem image_mul : (s * t).image (f : α → β) = s.image f * t.image f :=
+  image_image₂_distrib <| map_mul f
+#align finset.image_mul Finset.image_mul
+#align finset.image_add Finset.image_add
+
+/-- The singleton operation as a `MulHom`. -/
+@[to_additive "The singleton operation as an `add_hom`."]
```
Comment (jcommelin) [inline Mathlib/Data/Finset/Pointwise.lean:481]:
```suggestion
@[to_additive "The singleton operation as an `AddHom`."]
```

File: Mathlib/Data/Finset/Pointwise.lean [hunk 3]
```diff
@@ -0,0 +1,2010 @@
...
+
+section Div
+
+variable [DecidableEq α] [Div α] {s s₁ s₂ t t₁ t₂ u : Finset α} {a b : α}
+
+/-- The pointwise division of sfinets `s / t` is defined as `{x / y | x ∈ s, y ∈ t}` in locale
+`pointwise`. -/
+@[to_additive
+      "The pointwise subtraction of finsets `s - t` is defined as `{x - y | x ∈ s, y ∈ t}`
+      in locale `pointwise`."]
```
Comment (jcommelin) [inline Mathlib/Data/Finset/Pointwise.lean:521]:
```suggestion
      in locale `Pointwise`."]
```

File: Mathlib/Data/Finset/Pointwise.lean [hunk 4]
```diff
@@ -0,0 +1,2010 @@
...
+multiplication/division!) of a `Finset`. See note [pointwise nat action]. -/
+@[to_additive]
+protected def zpow [One α] [Mul α] [Inv α] : Pow (Finset α) ℤ :=
+  ⟨fun s n => zpowRec n s⟩
+#align finset.has_zpow Finset.zpow
+
+scoped[Pointwise] attribute [instance] Finset.nsmul Finset.npow Finset.zsmul Finset.zpow
+
+/-- `Finset α` is a `Semigroup` under pointwise operations if `α` is. -/
+@[to_additive "`Finset α` is an `add_semigroup` under pointwise operations if `α` is. "]
```
Comment (jcommelin) [inline Mathlib/Data/Finset/Pointwise.lean:731]:
```suggestion
@[to_additive "`Finset α` is an `AddSemigroup` under pointwise operations if `α` is. "]
```

File: Mathlib/Data/Finset/Pointwise.lean [hunk 5]
```diff
@@ -0,0 +1,2010 @@
...
+protected def semigroup [Semigroup α] : Semigroup (Finset α) :=
+  coe_injective.semigroup _ coe_mul
+#align finset.semigroup Finset.semigroup
+#align finset.add_semigroup Finset.addSemigroup
+
+/-- `Finset α` is a `CommSemigroup` under pointwise operations if `α` is. -/
+@[to_additive "`Finset α` is an `add_comm_semigroup` under pointwise operations if `α` is. "]
```
Comment (jcommelin) [inline Mathlib/Data/Finset/Pointwise.lean:738]:
```suggestion
@[to_additive "`Finset α` is an `AddCommSemigroup` under pointwise operations if `α` is. "]
```

File: Mathlib/Data/Finset/Pointwise.lean [hunk 6]
```diff
@@ -0,0 +1,2010 @@
...
+  coe_injective.commSemigroup _ coe_mul
+#align finset.comm_semigroup Finset.commSemigroup
+#align finset.add_comm_semigroup Finset.addCommSemigroup
+
+section MulOneClass
+
+variable [MulOneClass α]
+
+/-- `Finset α` is a `MulOneClass` under pointwise operations if `α` is. -/
+@[to_additive "`Finset α` is an `add_zero_class` under pointwise operations if `α` is."]
```
Comment (jcommelin) [inline Mathlib/Data/Finset/Pointwise.lean:749]:
```suggestion
@[to_additive "`Finset α` is an `AddZeroClass` under pointwise operations if `α` is."]
```

File: Mathlib/Data/Finset/Pointwise.lean [hunk 7]
```diff
@@ -0,0 +1,2010 @@
...
+#align finset.subset_add_left Finset.subset_add_left
+
+@[to_additive]
+theorem subset_mul_right {s : Finset α} (t : Finset α) (hs : (1 : α) ∈ s) : t ⊆ s * t := fun a ha =>
+  mem_mul.2 ⟨1, a, hs, ha, one_mul _⟩
+#align finset.subset_mul_right Finset.subset_mul_right
+#align finset.subset_add_right Finset.subset_add_right
+
+/-- The singleton operation as a `monoid_hom`. -/
+@[to_additive "The singleton operation as an `add_monoid_hom`."]
```
Comment (jcommelin) [inline Mathlib/Data/Finset/Pointwise.lean:771]:
```suggestion
/-- The singleton operation as a `MonoidHom`. -/
@[to_additive "The singleton operation as an `AddMonoidHom`."]
```

File: Mathlib/Data/Finset/Pointwise.lean [hunk 8]
```diff
@@ -0,0 +1,2010 @@
...
+#align finset.coe_singleton_add_monoid_hom Finset.coe_singletonAddMonoidHom
+
+@[to_additive (attr := simp)]
+theorem singletonMonoidHom_apply (a : α) : singletonMonoidHom a = {a} :=
+  rfl
+#align finset.singleton_monoid_hom_apply Finset.singletonMonoidHom_apply
+#align finset.singleton_add_monoid_hom_apply Finset.singletonAddMonoidHom_apply
+
+/-- The coercion from `Finset` to `set` as a `monoid_hom`. -/
+@[to_additive "The coercion from `Finset` to `set` as an `add_monoid_hom`."]
```
Comment (jcommelin) [inline Mathlib/Data/Finset/Pointwise.lean:790]:
```suggestion
/-- The coercion from `Finset` to `Set` as a `MonoidHom`. -/
@[to_additive "The coercion from `Finset` to `Set` as an `AddMonoidHom`."]
```

File: Mathlib/Data/Finset/Pointwise.lean [hunk 9]
```diff
@@ -0,0 +1,2010 @@
...
+#align finset.coe_coe_add_monoid_hom Finset.coe_coeAddMonoidHom
+
+@[to_additive (attr := simp)]
+theorem coeMonoidHom_apply (s : Finset α) : coeMonoidHom s = s :=
+  rfl
+#align finset.coe_monoid_hom_apply Finset.coeMonoidHom_apply
+#align finset.coe_add_monoid_hom_apply Finset.coeAddMonoidHom_apply
+
+/-- Lift a `monoid_hom` to `Finset` via `image`. -/
+@[to_additive "Lift an `add_monoid_hom` to `Finset` via `image`", simps]
```
Comment (jcommelin) [inline Mathlib/Data/Finset/Pointwise.lean:811]:
```suggestion
/-- Lift a `MonoidHom` to `Finset` via `image`. -/
@[to_additive "Lift an `AddMonoidHom` to `Finset` via `image`", simps]
```

File: Mathlib/Data/Finset/Pointwise.lean [hunk 10]
```diff
@@ -0,0 +1,2010 @@
...
+@[to_additive (attr := simp, norm_cast)]
+theorem coe_pow (s : Finset α) (n : ℕ) : ↑(s ^ n) = (s: Set α) ^ n  := by
+  change ↑(npowRec n s) = (s: Set α) ^ n
+  induction' n with n ih
+  · simp [npowRec, pow_zero, coe_one]
+  · simp [npowRec, pow_succ, coe_mul, ih]
+#align finset.coe_pow Finset.coe_pow
+
+/-- `Finset α` is a `Monoid` under pointwise operations if `α` is. -/
+@[to_additive "`Finset α` is an `add_monoid` under pointwise operations if `α` is. "]
```
Comment (jcommelin) [inline Mathlib/Data/Finset/Pointwise.lean:837]:
```suggestion
@[to_additive "`Finset α` is an `AddMonoid` under pointwise operations if `α` is. "]
```

**Comment by arienmalec** [comment]
I'll just run `fix-comments.sh` here -- may have forgotten this step.

**Comment by jcommelin** [comment]
@arienmalec unfortunately `fix-comments.py` doesn't know how to fix the docstrings passed to `to_additive`. So please fix those manually.

**Comment by arienmalec** [comment]
> @arienmalec unfortunately `fix-comments.py` doesn't know how to fix the docstrings passed to `to_additive`. So please fix those manually.

Ah, I see it now. Done.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**